cpp/src/jni/parquet/adapter.h

```diff
@@ -0,0 +1,159 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#ifndef ARROW_PARQUET_CONVERTER_H
+#define ARROW_PARQUET_CONVERTER_H
+
+#include <cstdint>
+#include <memory>
+#include <vector>
+
+#include "arrow/io/interfaces.h"
+#include "arrow/memory_pool.h"
+#include "arrow/record_batch.h"
+#include "arrow/status.h"
+#include "arrow/type.h"
+#include "arrow/util/visibility.h"
+#include "parquet/properties.h"
+
+namespace jni {
+
+namespace parquet {
+
+namespace adapters {
+
+using MemoryPool = arrow::MemoryPool;
+using OutputStream = arrow::io::OutputStream;
+using RandomAccessFile = arrow::io::RandomAccessFile;
+using RecordBatch = arrow::RecordBatch;
+using Schema = arrow::Schema;
+using Status = arrow::Status;
+
+/// \class ParquetFileReader
+/// \brief Read an Arrow RecordBatch from an PARQUET file.
+class ARROW_EXPORT ParquetFileReader {
+ public:
+  ~ParquetFileReader();
+
+  /// \brief Creates a new PARQUET reader.
+  ///
+  /// \param[in] file the data source
+  /// \param[in] pool a MemoryPool to use for buffer allocations
+  /// \param[in] properties ArrowReaderProperties
+  /// \param[out] reader the returned reader object
+  /// \return Status
+  static Status Open(std::shared_ptr<RandomAccessFile>& file, MemoryPool* pool,
+                     ::parquet::ArrowReaderProperties properties,
+                     std::unique_ptr<ParquetFileReader>* reader);
+
+  /// \brief Creates a new PARQUET reader.
+  ///
+  /// \param[in] file the data source
+  /// \param[in] pool a MemoryPool to use for buffer allocations
+  /// \param[out] reader the returned reader object
+  /// \return Status
+  static Status Open(std::shared_ptr<RandomAccessFile>& file, MemoryPool* pool,
```

Comment:
it would be better to use Result<std::unique_ptr<...>> instead of the out parameter unless there is a good reason not to.